mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp

```diff
@@ -346,9 +348,249 @@ propagateUnmergeVectorization(T &&dimAndLength,
+          dimensionMap[d] = d;
+        }
+        break;
+      case TransformType::AddDim:
+        dimToDelete.insert(upperDims[0]);
+        break;
+      case TransformType::Embed: {
+        // Sort the parameters
+        auto &&zip = llvm::zip(params, upperDims);
+        SmallVector<std::tuple<int64_t, uint32_t>> data(zip.begin(), zip.end());
+        std::sort(data.begin(), data.end());
+
+        // Verify that the Embed is an Unmerge operation and
+        // at the same time create the sorted unmerge params
+        bool maybeUnmerge = true;
+        SmallVector<int64_t> unmergeParams;
+        SmallVector<uint32_t> unmergeDims;
+        for (size_t i = 0; i < data.size() - 1; i++) {
+          auto paramI = std::get<0>(data[i]);
+          auto paramI1 = std::get<0>(data[i + 1]);
+          auto dim = std::get<1>(data[i]);
+          auto unmergeParam = paramI1 / paramI;
+          if (unmergeParam * paramI != paramI1 || paramI == paramI1) {
+            maybeUnmerge = false;
+            break;
+          }
+          unmergeParams.push_back(unmergeParam);
+          unmergeDims.push_back(dim);
+        }
+
+        // We are not done yet. To be sure this is an unmerge
+        // the faster parameter needs to be 1 and the upper bounds
+        // have to be set
+        auto fasterParam = std::get<0>(data[0]);
+        if (maybeUnmerge && fasterParam == 1 && upperBounds.size()) {
+          auto slowerDim = std::get<1>(data.back());
+          unmergeParams.push_back(upperBounds[slowerDim]);
+          unmergeDims.push_back(slowerDim);
+
+          // Shuffle the sorted unmerge params
+          auto &&zip = llvm::zip(unmergeDims, unmergeParams);
+          SmallVector<std::tuple<uint32_t, int64_t>> data(zip.begin(),
+                                                          zip.end());
+          std::sort(data.begin(), data.end());
+
+          SmallVector<int64_t> unmergeParamsReshuffled;
+          for (auto pair : data) {
+            unmergeParamsReshuffled.push_back(std::get<1>(pair));
+          }
+
+          // Now we can use the parameters as unmerge parameters and see if
+          // there are contiguous dimensions to be folded together.
+          scanForContiguousDimensions(upperDims, unmergeParamsReshuffled,
+                                      dimensionMap, dimToDelete, dimToMerge,
+                                      mergeInfoMap, result);
+        }
+        break;
+      }
+
+      case TransformType::PassThrough:
+      case TransformType::Pad:
+      case TransformType::Broadcast:
+        // Map the lower dimensions to the dimensions originally
+        // used (i.e., upper dimensions)
+        for (auto pair : llvm::zip(upperDims, lowerDims)) {
+          auto u = std::get<0>(pair);
+          auto l = std::get<1>(pair);
+          dimensionMap[l] = u;
+        }
+        break;
+      case TransformType::Slice:
+      case TransformType::Unfold: // TODO: remove Unfold
```

Comment:
For this PR, make this be the Merge case. But, collect up all the Unfold{} transforms and assert they all got recorded as contiguous groups at the end.